src/HealthChecks/HealthChecks/test/HealthCheckPublisherHostedServiceTest.cs

```diff
@@ -385,6 +486,103 @@ public async Task RunAsync_CanFilterHealthChecks()
+    {
+        // Arrange
+        const string HealthyMessage = "Everything is A-OK";
+
+        var unblockDelayedCheck = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        var publishers = new TestPublisher[]
+        {
+                new TestPublisher(),
+                new TestPublisher(),
+        };
+
+        var service = CreateService(
+            publishers,
+            configurePublisherOptions: (options) =>
+            {
+                options.Predicate = (r) => r.Name.Contains("Delay");
+            },
+            configureBuilder: b =>
+            {
+                b.Add(
+                    new(
+                        name: "CheckDefault",
+                        instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                        failureStatus: null,
+                        tags: null));
+
+                b.Add(
+                    new HealthCheckRegistration(
+                        name: "CheckDelay2Period18",
+                        instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                        failureStatus: null,
+                        tags: null)
+                    {
+                        Delay = TimeSpan.FromSeconds(2),
+                        Period = TimeSpan.FromSeconds(18)
+                    });
+
+                b.Add(
+                    new HealthCheckRegistration(
+                        name: "CheckDelay7Period11",
+                        instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                        failureStatus: null,
+                        tags: null,
+                        timeout: default)
+                    {
+                        Delay = TimeSpan.FromSeconds(7),
+                        Period = TimeSpan.FromSeconds(11)
+                    });
+
+                b.Add(
+                   new HealthCheckRegistration(
+                       name: "CheckDelay9Period5",
+                       instance: new DelegateHealthCheck(_ =>
+                       {
+                           unblockDelayedCheck.TrySetResult(null); // Unblock last delayed check
+                           return Task.FromResult(HealthCheckResult.Healthy(HealthyMessage));
+                       }),
+                       failureStatus: null,
+                       tags: null,
+                       timeout: default)
+                   {
+                       Delay = TimeSpan.FromSeconds(9),
+                       Period = TimeSpan.FromSeconds(5)
+                   });
+            });
+
+        try
+        {
+            await service.StartAsync();
+
+            // Act
+            await RunServiceAsync(service).TimeoutAfter(TimeSpan.FromSeconds(20));
+
+            await unblockDelayedCheck.Task;
```

Comment:
```suggestion
            await unblockDelayedCheck.Task.TimeoutAfter(TimeSpan.FromSeconds(10));
```

Since this test isn't verifying `RunServiceAsync` blocks, we can just set the default `Delay` to `TimeSpan.Zero` and trust it will finish before the .9 second one. We should time it out though just in case  the .9 second delay registration does fail to run for some reason.